File: telemetry-go/telemetry.go
package telemetry

import (
	"context"
	"net"
	"strings"

	"github.com/the-protobuf-project/telemetry/telemetry-go/internal/foxglove"
	"github.com/the-protobuf-project/telemetry/telemetry-go/internal/logging"
	"github.com/the-protobuf-project/telemetry/telemetry-go/internal/metrics"
	"github.com/the-protobuf-project/telemetry/telemetry-go/internal/profiling"
	"github.com/the-protobuf-project/telemetry/telemetry-go/internal/telemetry"
	"github.com/the-protobuf-project/telemetry/telemetry-go/internal/tracing"
	"github.com/the-protobuf-project/telemetry/telemetry-go/options"
)

// Span is a type alias for tracing.Span to avoid exposing internal packages
type Span = tracing.Span

// LogLevel is a type alias for options.LogLevel so modules can use telemetry.Level1, etc.
type LogLevel = options.LogLevel

// Re-export OpenTelemetry metric functions from internal metrics package
var WithAttributes = metrics.WithAttributes

// Re-export OpenTelemetry attribute creation functions from internal metrics package
var StringAttribute = metrics.StringAttribute

// Log level constants for modules.
//
//   - ModuleLevel_1 (Error)  — stable, production-ready module
//   - ModuleLevel_2 (Info)   — normal operation
//   - ModuleLevel_3 (Debug)  — active development, full observability
const (
	ModuleLevel_1 = options.ModuleLevel_1
	ModuleLevel_2 = options.ModuleLevel_2
	ModuleLevel_3 = options.ModuleLevel_3
)

// Telemetry is the main framework struct that provides access to all telemetry services.
// It supports both the legacy logging interface and the new unified OpenTelemetry-based telemetry.
type Telemetry struct {
	// Logger is the main logging client.
	Logger *logging.Logger
	// Metrics is the main metrics client.
	Metrics *metrics.Metrics
	// Tracing is the main tracing client.
	Tracing *tracing.Tracing
	// Profiler is the main profiler client.
	Profiler *profiling.Profiler

	// Unified OpenTelemetry-based telemetry
	telemetry *telemetry.Telemetry
	// Unified MCAP writer for both logs and metrics
	unifiedMcap *foxglove.UnifiedMcapWriter
	// Internal context for background operations
	ctx context.Context
}

// Builder provides a fluent API for configuring and creating a Telemetry instance.
type Builder struct {
	serviceOpts        options.ServiceOptions
	telemetryOpts options.TelemetryOptions
	configPath         string
	err                error
}

// New creates a new Telemetry Builder with sensible defaults.
// Use the builder methods to configure, then call Build() to create the Telemetry instance.
//
// Example:
//
//	p, err := telemetry.New().
//	    WithService("my-service", "1.0.0").
//	    WithConfig("config.yaml").
// New creates a Builder initialized with auto-discovered telemetry and service configuration, using defaults when configuration is unavailable.
func New() *Builder {
	// Auto-discover and load config on creation
	telemetryOpts, serviceOpts, _ := options.LoadConfigWithDefaults("")
	return &Builder{
		serviceOpts:        *serviceOpts,
		telemetryOpts: *telemetryOpts,
	}
}

// WithConfig loads configuration from a specific file (YAML, JSON, or TOML).
// Use this to override the auto-discovered config.
// Environment variables with TELEMETRY_ prefix override file values.
func (b *Builder) WithConfig(configPath string) *Builder {
	if b.err != nil {
		return b
	}
	if configPath == "" {
		return b // Already loaded via auto-discovery
	}
	b.configPath = configPath
	telemetryOpts, serviceOpts, err := options.LoadConfigWithDefaults(configPath)
	if err != nil {
		b.err = err
		return b
	}
	b.telemetryOpts = *telemetryOpts
	b.serviceOpts = *serviceOpts
	return b
}

// WithService sets the service name and version.
// When this is called, it clears any service-level configuration from the config file
// to avoid collisions between config file and code-level settings.
func (b *Builder) WithService(name, version string) *Builder {
	if b.err != nil {
		return b
	}
	// Reset service options to defaults, ignoring config file values
	b.serviceOpts = options.ServiceOptions{
		Name:        name,
		Version:     version,
		Description: "",
		Environment: options.Development,
		Labels:      make(map[string]string),
	}
	return b
}

// WithDescription sets the service description.
func (b *Builder) WithDescription(description string) *Builder {
	if b.err != nil {
		return b
	}
	b.serviceOpts.Description = description
	return b
}

// WithEnvironment sets the deployment environment.
func (b *Builder) WithEnvironment(env options.Environment) *Builder {
	if b.err != nil {
		return b
	}
	b.serviceOpts.Environment = env
	return b
}

// WithLogLevel sets the log level for this service/module.
// This acts as the code-level default. It can be overridden by the config file
// via [logging.modules.<service-name>] or env vars (TELEMETRY_LOGGING_MODULES_<NAME>_LEVEL).
//
// Priority chain (highest to lowest):
//
//	env var > TOML per-module override > WithLogLevel() > environment-based default
//
// Example:
//
//	p, err := telemetry.New().
//	    WithService("vision", "1.0.0").
//	    WithLogLevel(telemetry.ModuleLevel_3). // "I'm a level 3 module"
//	    Build()
func (b *Builder) WithLogLevel(level LogLevel) *Builder {
	if b.err != nil {
		return b
	}
	// Store as a per-module override keyed by service name.
	// This will be checked in resolveLogLevel via the Modules map.
	// If the TOML also has a module override for this name, the TOML wins
	// because config is loaded first and WithLogLevel is applied only
	// when no config override exists.
	if b.telemetryOpts.Logging.Modules == nil {
		b.telemetryOpts.Logging.Modules = make(map[string]options.ModuleOptions)
	}
	// Only set if config didn't already provide a per-module override for this service
	if _, exists := b.telemetryOpts.Logging.Modules[b.serviceOpts.Name]; !exists {
		b.telemetryOpts.Logging.Modules[b.serviceOpts.Name] = options.ModuleOptions{
			Level: level,
		}
	}
	return b
}

// WithOTLP configures the OTLP endpoint. Auto-detects if local or remote
// and configures TLS accordingly.
func (b *Builder) WithOTLP(host string, port int) *Builder {
	if b.err != nil {
		return b
	}
	b.telemetryOpts.Telemetry.OTLP.Host = host //nolint:staticcheck // Deprecated but kept for backward compatibility
	b.telemetryOpts.Telemetry.OTLP.Port = port
	b.telemetryOpts.Telemetry.OTLP.Enabled = true

	// Auto-configure based on host
	b.telemetryOpts.Telemetry.OTLP.Secure = !isLocalHost(host)
	return b
}

// WithOTLPHeaders sets custom headers for OTLP requests (e.g., Authorization).
func (b *Builder) WithOTLPHeaders(headers map[string]string) *Builder {
	if b.err != nil {
		return b
	}
	b.telemetryOpts.Telemetry.OTLP.Headers = headers
	return b
}

// WithMCAP enables MCAP file logging at the specified path.
func (b *Builder) WithMCAP(path string) *Builder {
	if b.err != nil {
		return b
	}
	b.telemetryOpts.Foxglove.Enabled = true
	b.telemetryOpts.Foxglove.McapPath = path
	return b
}

// WithProfiling enables continuous profiling with Pyroscope.
func (b *Builder) WithProfiling(serverAddress string) *Builder {
	if b.err != nil {
		return b
	}
	b.telemetryOpts.Profiling.Enabled = true
	b.telemetryOpts.Profiling.ServerAddress = serverAddress
	return b
}

// WithTracing enables distributed tracing.
func (b *Builder) WithTracing() *Builder {
	if b.err != nil {
		return b
	}
	b.telemetryOpts.Tracing.Enabled = true
	b.telemetryOpts.Telemetry.Tracing.Enabled = true
	return b
}

// WithLabels sets global labels that are added to all telemetry data.
// Use this for constant identifiers like robot.id, device.id, fleet.id, etc.
// These labels will appear on all logs, metrics, and traces.
//
// Example:
//
//	p, err := telemetry.New().
//	    WithService("robot-controller", "1.0.0").
//	    WithLabels(map[string]string{
//	        "robot.id":    "robot-001",
//	        "fleet.id":    "fleet-alpha",
//	        "location.id": "warehouse-1",
//	    }).
//	    Build()
func (b *Builder) WithLabels(labels map[string]string) *Builder {
	if b.err != nil {
		return b
	}
	if b.serviceOpts.Labels == nil {
		b.serviceOpts.Labels = make(map[string]string)
	}
	for k, v := range labels {
		b.serviceOpts.Labels[k] = v
	}
	return b
}

// WithLabel sets a single global label.
// Convenience method for adding one label at a time.
func (b *Builder) WithLabel(key, value string) *Builder {
	if b.err != nil {
		return b
	}
	if b.serviceOpts.Labels == nil {
		b.serviceOpts.Labels = make(map[string]string)
	}
	b.serviceOpts.Labels[key] = value
	return b
}

// Build creates the Telemetry instance with the configured options.
func (b *Builder) Build() (*Telemetry, error) {
	if b.err != nil {
		return nil, b.err
	}

	// Auto-configure OTLP settings based on host
	autoConfigureOTLP(&b.telemetryOpts.Telemetry.OTLP)

	ctx := context.Background()

	// Initialize unified telemetry service
	tel, err := telemetry.New(ctx, b.serviceOpts, b.telemetryOpts.Telemetry)
	if err != nil {
		return nil, err
	}

	// Initialize unified MCAP writer if Foxglove is enabled
	var unifiedMcap *foxglove.UnifiedMcapWriter
	if b.telemetryOpts.Foxglove.Enabled && b.telemetryOpts.Foxglove.McapPath != "" {
		unifiedMcap, err = foxglove.NewUnifiedMcapWriter(b.serviceOpts, b.telemetryOpts.Foxglove)
		if err != nil {
			return nil, err
		}
	}

	p := &Telemetry{
		ctx:         ctx,
		telemetry:   tel,
		unifiedMcap: unifiedMcap,
		Logger:      logging.NewLogger(b.serviceOpts, b.telemetryOpts.Logging, unifiedMcap, tel.GetLogger()),
		Metrics:     metrics.NewMetrics(b.serviceOpts, unifiedMcap, tel.GetMetrics()),
		Tracing:     tracing.NewTracing(b.serviceOpts, b.telemetryOpts.Tracing, unifiedMcap, tel.GetTracer()),
		Profiler:    profiling.NewProfiler(b.serviceOpts, b.telemetryOpts.Profiling, unifiedMcap),
	}

	return p, nil
}

// isLocalHost reports whether host is empty, localhost, a loopback address, or a private IP address.
func isLocalHost(host string) bool {
	host = strings.ToLower(host)
	if host == "localhost" || host == "" {
		return true
	}

	ip := net.ParseIP(host)
	if ip == nil {
		// Not an IP, check if it's a domain
		return false
	}

	// Check for loopback (127.x.x.x)
	if ip.IsLoopback() {
		return true
	}

	// Check for private IPs (10.x.x.x, 172.16-31.x.x, 192.168.x.x)
	if ip.IsPrivate() {
		return true
	}

	return false
}

// autoConfigureOTLP enables OTLP and applies transport defaults based on the configured endpoint or host.
// autoConfigureOTLP enables OTLP when an endpoint or host is configured and applies TLS and HTTP defaults for remote hosts on ports 443 and 4318.
func autoConfigureOTLP(otlp *options.OTLPOptions) {
	// Auto-enable OTLP when an endpoint is configured
	if !otlp.Enabled && otlp.Endpoint != "" {
		otlp.Enabled = true
	}

	//nolint:staticcheck // Deprecated but kept for backward compatibility
	if otlp.Host == "" {
		return
	}

	// Auto-enable OTLP when a host is configured (deprecated path)
	if !otlp.Enabled {
		otlp.Enabled = true
	}

	isLocal := isLocalHost(otlp.Host) //nolint:staticcheck // Deprecated but kept for backward compatibility

	// Auto-set secure based on host (unless explicitly set)
	if !otlp.Secure && !isLocal {
		// Remote host on standard OTLP ports - check if it needs TLS
		switch otlp.Port {
		case 443, 4318:
			otlp.Secure = true
			otlp.UseHTTP = true // Port 443 typically needs HTTP
		case 4317:
			// Standard gRPC port - may or may not need TLS
			// Keep as-is, user should configure if needed
		}
	}
}

// NewLegacy creates a new Telemetry instance using the legacy API (for backward compatibility).
// NewLegacy creates telemetry clients from an explicit context and configuration.
// NewLegacy creates telemetry clients from an explicit context and configuration.
// Deprecated: Use New().WithConfig().Build() instead.
func NewLegacy(ctx context.Context, serviceOpts options.ServiceOptions, opts options.TelemetryOptions) (*Telemetry, error) {
	// Auto-configure OTLP settings
	autoConfigureOTLP(&opts.Telemetry.OTLP)

	// Initialize unified telemetry service
	tel, err := telemetry.New(ctx, serviceOpts, opts.Telemetry)
	if err != nil {
		return nil, err
	}

	// Initialize unified MCAP writer if Foxglove is enabled
	var unifiedMcap *foxglove.UnifiedMcapWriter
	if opts.Foxglove.Enabled && opts.Foxglove.McapPath != "" {
		unifiedMcap, err = foxglove.NewUnifiedMcapWriter(serviceOpts, opts.Foxglove)
		if err != nil {
			return nil, err
		}
	}

	p := &Telemetry{
		ctx:         ctx,
		telemetry:   tel,
		unifiedMcap: unifiedMcap,
		Logger:      logging.NewLogger(serviceOpts, opts.Logging, unifiedMcap, tel.GetLogger()),
		Metrics:     metrics.NewMetrics(serviceOpts, unifiedMcap, tel.GetMetrics()),
		Tracing:     tracing.NewTracing(serviceOpts, opts.Tracing, unifiedMcap, tel.GetTracer()),
		Profiler:    profiling.NewProfiler(serviceOpts, opts.Profiling, unifiedMcap),
	}

	return p, nil
}

// Close gracefully shuts down all telemetry services.
// Uses the internal context created during Build().
func (p *Telemetry) Close() error {
	// Stop profiler first to flush remaining data
	if p.Profiler != nil {
		if err := p.Profiler.Stop(); err != nil {
			// Log error but continue with shutdown
			if p.Logger != nil {
				p.Logger.Warn("Failed to stop profiler", map[string]interface{}{"error": err.Error()})
			}
		}
	}

	// Close unified MCAP writer first (before logger tries to log about it)
	if p.unifiedMcap != nil {
		_ = p.unifiedMcap.Close() // Ignore error during shutdown
	}

	// Close metrics (no-op since unified writer is already closed)
	if p.Metrics != nil {
		_ = p.Metrics.Close() // Ignore error during shutdown
	}

	// Close logger (no-op since unified writer is already closed)
	if p.Logger != nil {
		_ = p.Logger.Close() // Ignore error during shutdown
	}

	// Close tracing (no-op since unified writer is already closed)
	if p.Tracing != nil {
		_ = p.Tracing.Close() // Ignore error during shutdown
	}

	if p.telemetry != nil {
		ctx := p.ctx
		if ctx == nil {
			ctx = context.Background()
		}
		return p.telemetry.Shutdown(ctx)
	}
	return nil
}

File: telemetry-py/telemetry/telemetry.py
from typing import Optional, Dict

from .options import (
    ServiceOptions,
    TelemetryOptions,
    Environment,
    LogLevel,
    ModuleOptions,
    from_config,
)
from ._private.logging import TelemetryLogger
from ._private.metrics import (
    TelemetryMetrics,
    set_current_telemetry_metrics,
    reset_current_telemetry_metrics,
)
from ._private.tracing import (
    TelemetryTracing,
    set_current_telemetry,
    reset_current_telemetry,
)
from ._private.foxglove import UnifiedMcapWriter


class TelemetryBuilder:
    """Builder for creating Telemetry instances with fluent API.

    Configuration Priority (Lowest to Highest):
    1. Defaults (lowest priority)
    2. Config file (telemetry.toml / telemetry.yaml / telemetry.json)
    3. Environment variables (TELEMETRY_*)
    4. Code-based (builder methods) - highest priority

    Example:
        telemetry = Telemetry.new() \\
            .with_service("my-service", "1.0.0") \\
            .environment(Environment.PRODUCTION) \\
            .with_otlp("localhost", 4317) \\
            .build()
    """

    def __init__(self):
        self._config_path: Optional[str] = None
        self._name: Optional[str] = None
        self._version: Optional[str] = None
        self._description: Optional[str] = None
        self._environment: Optional[Environment] = None
        self._labels: Dict[str, str] = {}
        self._otlp_endpoint: Optional[str] = None
        self._otlp_auth_token: Optional[str] = None
        self._otlp_secure: Optional[bool] = None
        self._otlp_use_http: Optional[bool] = None
        self._mcap_path: Optional[str] = None
        self._tracing_enabled: bool = False
        self._log_level: Optional[LogLevel] = None
        self._service_from_code: bool = False

    def with_config(self, config_path: str) -> "TelemetryBuilder":
        """Load configuration from a specific file path."""
        self._config_path = config_path
        return self

    def with_service(self, name: str, version: str) -> "TelemetryBuilder":
        """Set service name and version.

        When this is called, it indicates the user wants to configure service via code,
        so we'll clear any service-level configuration from the config file to avoid collisions.
        """
        self._name = name
        self._version = version
        # Mark that service should be configured via code only
        self._service_from_code = True
        return self

    def description(self, desc: str) -> "TelemetryBuilder":
        """Set service description."""
        self._description = desc
        return self

    def environment(self, env: Environment) -> "TelemetryBuilder":
        """Set deployment environment."""
        self._environment = env
        return self

    def with_label(self, key: str, value: str) -> "TelemetryBuilder":
        """Add a custom label to all telemetry."""
        self._labels[key] = value
        return self

    def with_labels(self, labels: Dict[str, str]) -> "TelemetryBuilder":
        """Add multiple custom labels to all telemetry."""
        self._labels.update(labels)
        return self

    def with_otlp(self, host: str, port: int) -> "TelemetryBuilder":
        """Enable OTLP export to the specified endpoint."""
        self._otlp_endpoint = f"{host}:{port}"
        return self

    def with_otlp_endpoint(self, endpoint: str) -> "TelemetryBuilder":
        """Enable OTLP export to the specified endpoint string."""
        self._otlp_endpoint = endpoint
        return self

    def with_otlp_auth(self, token: str) -> "TelemetryBuilder":
        """Set OTLP authentication token."""
        self._otlp_auth_token = token
        return self

    def with_otlp_secure(self, secure: bool = True) -> "TelemetryBuilder":
        """Enable/disable TLS for OTLP connection."""
        self._otlp_secure = secure
        return self

    def with_otlp_http(self, use_http: bool = True) -> "TelemetryBuilder":
        """Use HTTP instead of gRPC for OTLP."""
        self._otlp_use_http = use_http
        return self

    def with_mcap(self, path: str) -> "TelemetryBuilder":
        """Enable MCAP recording to the specified file path."""
        self._mcap_path = path
        return self

    def with_log_level(self, level: LogLevel) -> "TelemetryBuilder":
        """Set the log level for this service/module.

        This acts as the code-level default. It can be overridden by the
        config file via [logging.modules.<service-name>] or env vars.

        Priority chain (highest to lowest):
            env var > TOML per-module override > with_log_level() > environment default

        Example:
            telemetry = Telemetry.new() \\
                .with_service("vision", "1.0.0") \\
                .with_log_level(LogLevel.MODULE_LEVEL_3) \\
                .build()
        """
        self._log_level = level
        return self

    def with_tracing(self) -> "TelemetryBuilder":
        """Enable distributed tracing."""
        self._tracing_enabled = True
        return self

    def build(self) -> "Telemetry":
        """Build and return the Telemetry instance."""
        # Load config from file (auto-discovery or specified path)
        service_opts, telemetry_opts = from_config(self._config_path)

        # If with_service was called, ignore service-level config from file
        if self._service_from_code:
            service_opts = ServiceOptions(
                name=self._name or "telemetry-service",
                version=self._version or "1.0.0",
                description=self._description or "",
                environment=self._environment or Environment.DEVELOPMENT,
                labels=dict(self._labels),  # Copy builder labels
            )
        else:
            # Override with builder values (highest priority)
            if self._name:
                service_opts.name = self._name
            if self._version:
                service_opts.version = self._version
            if self._description:
                service_opts.description = self._description
            if self._environment:
                service_opts.environment = self._environment

            # Merge labels from builder with config
            if self._labels:
                service_opts.labels.update(self._labels)

        # Configure OTLP if specified via builder
        if self._otlp_endpoint:
            telemetry_opts.telemetry.otlp.enabled = True
            telemetry_opts.telemetry.otlp.endpoint = self._otlp_endpoint

        if self._otlp_auth_token:
            telemetry_opts.telemetry.otlp.auth_token = self._otlp_auth_token

        if self._otlp_secure is not None:
            telemetry_opts.telemetry.otlp.secure = self._otlp_secure

        if self._otlp_use_http is not None:
            telemetry_opts.telemetry.otlp.use_http = self._otlp_use_http

        # Apply code-level log level (only if config didn't already set a per-module override)
        if self._log_level is not None and self._name:
            modules = telemetry_opts.telemetry.logging.modules
            if self._name not in modules:
                modules[self._name] = ModuleOptions(level=self._log_level)

        # Enable tracing if requested
        if self._tracing_enabled:
            telemetry_opts.telemetry.tracing.enabled = True
            telemetry_opts.telemetry.otlp.enabled = True

        # Configure MCAP if specified
        if self._mcap_path:
            telemetry_opts.foxglove.enabled = True
            telemetry_opts.foxglove.mcap_path = self._mcap_path

        return Telemetry(service_opts, telemetry_opts)


class Telemetry:
    """
    Main Telemetry framework class providing unified observability.

    Integrates:
    - Logging (logbook + OpenTelemetry + MCAP)
    - Metrics (OpenTelemetry + MCAP with Pydantic support)
    - Tracing (OpenTelemetry + MCAP with decorator support)

    Example using builder pattern:
        telemetry = Telemetry.new() \\
            .with_service("my-service", "1.0.0") \\
            .environment(Environment.PRODUCTION) \\
            .build()

    Example using direct instantiation:
        telemetry = Telemetry(service_opts, telemetry_opts)
    """

    @classmethod
    def new(cls) -> "TelemetryBuilder":
        """Create a new TelemetryBuilder for fluent configuration."""
        return TelemetryBuilder()

    def __init__(
        self, service_opts: ServiceOptions, telemetry_opts: TelemetryOptions
    ):
        self.service_opts = service_opts
        self.telemetry_opts = telemetry_opts

        # Initialize unified MCAP writer if enabled
        self.mcap_writer: Optional[UnifiedMcapWriter] = None
        if (
            telemetry_opts.foxglove.enabled
            and telemetry_opts.foxglove.mcap_path
        ):
            self.mcap_writer = UnifiedMcapWriter(
                mcap_path=telemetry_opts.foxglove.mcap_path,
                service_name=service_opts.name,
            )

        # Initialize logging
        self.logger = TelemetryLogger(
            service_opts=service_opts,
            logging_opts=telemetry_opts.telemetry.logging,
            otlp_opts=telemetry_opts.telemetry.otlp
            if telemetry_opts.telemetry.otlp.enabled
            else None,
            mcap_writer=self.mcap_writer,
        )

        # Initialize metrics
        self.metrics = TelemetryMetrics(
            service_opts=service_opts,
            metrics_opts=telemetry_opts.telemetry.metrics,
            otlp_opts=telemetry_opts.telemetry.otlp
            if telemetry_opts.telemetry.otlp.enabled
            else None,
            mcap_writer=self.mcap_writer,
        )

        # Initialize tracing
        self.tracing = TelemetryTracing(
            service_opts=service_opts,
            tracing_opts=telemetry_opts.telemetry.tracing,
            otlp_opts=telemetry_opts.telemetry.otlp
            if telemetry_opts.telemetry.otlp.enabled
            else None,
            mcap_writer=self.mcap_writer,
        )

    def __enter__(self):
        """Enter context manager"""
        self._telemetry_token = set_current_telemetry(self)
        self._metrics_token = set_current_telemetry_metrics(self)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Exit context manager and close resources"""
        reset_current_telemetry_metrics(self._metrics_token)
        reset_current_telemetry(self._telemetry_token)
        self.close()
        return False

    def close(self):
        """Close all Telemetry components and flush pending data"""
        # Close components in order
        self.tracing.close()
        self.metrics.close()
        self.logger.close()

        # Close MCAP writer last
        if self.mcap_writer:
            self.mcap_writer.close()
